### PR TITLE
Merge Automatic back into AuthenticationHandler

### PR DESCRIPTION
**Pull request by HaoK**
All AuthMiddlewares are not automatic by default (passive), but now again have the ability to flip into Automatic mode again

cc @lodejard @Tratcher @Praburaj @brentschmaltz @blowdart

### CODE REVIEW

File: src/Microsoft.AspNet.Authentication/ClaimsTransformationAuthenticationHandler.cs [hunk 1]
```diff
@@ -30,7 +30,7 @@ private void ApplyTransform(IAuthenticateContext context)
             {
                 // REVIEW: this cast seems really bad (missing interface way to get the result back out?)
                 var authContext = context as AuthenticateContext;
-                if (context != null)
+                if (authContext != null && authContext.Result?.Principal != null)
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Authentication/ClaimsTransformationAuthenticationHandler.cs]: authContext?.Result?.Principal

**Comment by PinpointTownes** [comment]
OH YEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAH! :+1:

File: src/Microsoft.AspNet.Authentication/AuthenticationOptions.cs [hunk 2]
```diff
@@ -27,6 +27,13 @@ public string AuthenticationScheme
         }
 
         /// <summary>
+        /// If true the authentication middleware alter the request user coming in and
+        /// alter 401 Unauthorized responses going out. If false the authentication middleware will only provide
+        /// identity and alter responses when explicitly indicated by the AuthenticationScheme.
+        /// </summary>
+        public bool AutomaticAuthentication { get; set; }
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication/AuthenticationOptions.cs]: Is it worth resurrecting `AuthenticationMode`? (with `Active` -> `Automatic` and `Passive` -> `Manual`)?

File: test/Microsoft.AspNet.Authentication.Test/AuthenticationHandlerFacts.cs [hunk 3]
```diff
@@ -79,9 +78,13 @@ protected override AuthenticationTicket AuthenticateCore()
 
         private class TestOptions : AuthenticationOptions { }
 
-        private class TestAutoOptions : AutomaticAuthenticationOptions { }
+        private class TestAutoOptions : AuthenticationOptions {
```
Comment (PinpointTownes) [inline test/Microsoft.AspNet.Authentication.Test/AuthenticationHandlerFacts.cs]: Yeah, K&R is back too! :smile:

Comment (davidfowl) [inline test/Microsoft.AspNet.Authentication.Test/AuthenticationHandlerFacts.cs]: lol

File: src/Microsoft.AspNet.Authentication/AuthenticationOptions.cs (hunk 2 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Authentication/AuthenticationOptions.cs]: I believe this was supposed to be a simplification of that enum.

Comment (HaoK) [inline src/Microsoft.AspNet.Authentication/AuthenticationOptions.cs]: Yah this is an attempt at more clarity.

File: test/Microsoft.AspNet.Authentication.Test/MicrosoftAccount/MicrosoftAccountMiddlewareTests.cs [hunk 4]
```diff
@@ -169,7 +169,10 @@ private static TestServer CreateServer(Action<MicrosoftAccountAuthenticationOpti
                         options.SignInScheme = "External";
                     });
                 });
-                app.UseCookieAuthentication(options => options.AuthenticationScheme = "External");
+                app.UseCookieAuthentication(options => {
```
Comment (PinpointTownes) [inline test/Microsoft.AspNet.Authentication.Test/MicrosoftAccount/MicrosoftAccountMiddlewareTests.cs]: We really need more K&R :smile:

Comment (HaoK) [inline test/Microsoft.AspNet.Authentication.Test/MicrosoftAccount/MicrosoftAccountMiddlewareTests.cs]: Yeah well its what I prefer honestly and what I naturally use :)

**Comment by HaoK** [comment]
updated with PR feedback

File: src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs [hunk 5]
```diff
@@ -72,6 +72,15 @@ internal AuthenticationOptions BaseOptions
             Response.OnSendingHeaders(OnSendingHeaderCallback, this);
 
             await InitializeCoreAsync();
+
+            if (BaseOptions.AutomaticAuthentication)
+            {
+                AuthenticationTicket ticket = await AuthenticateAsync();
+                if (ticket != null && ticket.Principal != null)
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs]: ticket?.Principal != null :smile:

File: src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs [hunk 6]
```diff
@@ -72,6 +72,15 @@ internal AuthenticationOptions BaseOptions
             Response.OnSendingHeaders(OnSendingHeaderCallback, this);
 
             await InitializeCoreAsync();
+
+            if (BaseOptions.AutomaticAuthentication)
+            {
+                AuthenticationTicket ticket = await AuthenticateAsync();
```
Comment (Praburaj) [inline src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs]: minor : Can we use `var`.

**Comment by HaoK** [comment]
Cleaned up AuthHandler more var's and ?'s

File: src/Microsoft.AspNet.Authentication/AuthenticationOptions.cs [hunk 7]
```diff
@@ -27,6 +27,13 @@ public string AuthenticationScheme
         }
 
         /// <summary>
+        /// If true the authentication middleware alter the request user coming in and
```
Comment (Praburaj) [inline src/Microsoft.AspNet.Authentication/AuthenticationOptions.cs]: Can we rephrase this a little bit something like.. authentication middleware `will` set the request's <see cref=Request.User> and handle 401 unauthorized responses.

**Comment by Praburaj** [comment]
Looks good to me. :shipit:

File: src/Microsoft.AspNet.Authentication/AuthenticationOptions.cs (hunk 7 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication/AuthenticationOptions.cs]: Filed https://github.com/aspnet/Security/issues/190 to track this general issue, as I'm pretty bad at updating docs in general, I'm sure they are pretty out of date now

**Comment by PinpointTownes** [comment]
Out of curiosity, why switching Cookies and Bearer to manual/passive by default?

Adding documentation on the `IApplicationBuilder` extensions might be useful to warn people that `app.UseCookieAuthentication`/`app.UseOAuthBearerAuthentication` won't do anything by default.

**Comment by HaoK** [comment]
@PinpointTownes The gist of it is basically if you only have one auth middleware, that one probably should be automatic, but when you have more than one, typically only one will be automatic and the rest will be passive.  This makes it explicit choosing which auth you want to be automatic

**Comment by HaoK** [comment]
Merged 14d1b467c69dd6f836f5683b6e1545b7f3f859ee

**Pull request closed without merging**